lib/puppet/util/autoload.rb

```diff
@@ -55,6 +55,27 @@ def changed?(name)
     # Load a single plugin by name.  We use 'load' here so we can reload a
     # given plugin.
     def load_file(name, env)
+      # $clean_loadpath     - remembers LOAD_PATH from where we started
+      # $env_loadpath[env]  - remembers the LOAD_PATH we want to apply for each environment
+      # $current_loadpath   - which LOAD_PATH we have set right now
+      # loadpath            - which LOAD_PATH we want to apply
+      # last_loadpath       - which LOAD_PATH was set when we started
+      $clean_loadpath ||= $LOAD_PATH.dup
+      $env_loadpath ||= {}
```

Comment:
Sorry, but I don't think it is a good approach to add global variables for this (or for anything else for that matter). Also, since it is a global cache of env to loadpath and there is no cache eviction this will leak memory. As I suggested earlier this should go into the Puppet context (or possibly in the environment object).